File: lib/actions.ts
'use server';

import { tryit } from 'radash';
import { archiveMatch, persistRoom } from './db';
import { Game } from './game';
import { gameStore, sanitizeCode } from './game-store';

export type ActionResult<T = void> = { ok: true; data?: T } | { ok: false; error: string };

const ok = <T>(data?: T): ActionResult<T> => ({ ok: true, data });
const err = (error: string): ActionResult<never> => ({ ok: false, error });

const guard = (game: Game | undefined): ActionResult | null =>
  game ? null : err('房间不存在');

/**
 * Mirror the latest game state to the persistence layer. When phase is
 * 'ended', also archive the finished match — archiveMatch is idempotent
 * via UNIQUE(code, started_at), so multiple calls (a player rejoining
 * the post-game scoreboard, a restart, a leave) collapse to a single
 * matches row. The live snapshot stays in `rooms` until the host
 * resets or the room empties out.
 */
function persist(game: Game): void {
  if (game.phase === 'ended' && game.startedAt != null) {
    const winner = game.players.find((p) => p.id === game.winnerId);
    archiveMatch({
      code: game.code,
      winnerId: game.winnerId ?? null,
      winnerName: winner?.name ?? null,
      playerCount: game.players.length,
      startedAt: game.startedAt,
      endedAt: Date.now(),
      log: game.log,
      participants: game.players.map((p) => ({
        playerId: p.id,
        name: p.name,
        isWinner: p.id === game.winnerId,
        isHost: p.id === game.hostId,
      })),
    });
  }
  persistRoom(game.toSnapshot());
}

const run = async (code: string, mutate: (game: Game) => void): Promise<ActionResult> => {
  const game = gameStore.get(sanitizeCode(code));
  const fail = guard(game);
  if (fail) return fail;
  const [error] = await tryit(async () => mutate(game!))();
  if (error) return err((error as Error).message);
  persist(game!);
  game!.notify();
  return ok();
};

/** Either creates the room and joins, or rejoins an existing one with this playerId. */
export async function joinOrCreateRoom(
  rawCode: string,
  playerId: string,
  name: string,
  asHost: boolean,
): Promise<ActionResult> {
  if (!rawCode || !playerId || !name) return err('请提供房间码、玩家与昵称');
  const code = sanitizeCode(rawCode);

  let game = gameStore.get(code);
  if (!game) {
    game = new Game(code, playerId);
    gameStore.set(code, game);
    game.addPlayer(playerId, name);
    persist(game);
    game.notify();
    return ok();
  }

  if (game.players.some((p) => p.id === playerId)) {
    game.addPlayer(playerId, name);
    persist(game);
    game.notify();
    return ok();
  }

  if (asHost) return err('房间已存在，请改为加入');
  if (game.phase !== 'waiting') return err('对局已开始，无法加入');
  if (game.players.length >= 4) return err('房间已满');
  if (game.hasPlayerName(name)) return err('昵称已被占用');

  game.addPlayer(playerId, name);
  persist(game);
  game.notify();
  return ok();
}

export async function startGame(code: string, playerId: string): Promise<ActionResult> {
  return run(code, (g) => g.start(playerId));
}

export async function drawTile(
  code: string,
  playerId: string,
  color: 'black' | 'white',
): Promise<ActionResult> {
  return run(code, (g) => g.drawTile(playerId, color));
}

export async function guessTile(
  code: string,
  playerId: string,
  targetPlayerId: string,
  tileId: string,
  /** null = guessing the joker ("-"). */
  number: number | null,
): Promise<ActionResult> {
  return run(code, (g) => { g.guess(playerId, targetPlayerId, tileId, number); });
}

export async function placeJoker(
  code: string,
  playerId: string,
  position: number,
): Promise<ActionResult> {
  return run(code, (g) => g.placeJoker(playerId, position));
}

export async function decideContinue(
  code: string,
  playerId: string,
  cont: boolean,
): Promise<ActionResult> {
  return run(code, (g) => g.decideContinue(playerId, cont));
}

export async function resetGame(code: string, playerId: string): Promise<ActionResult> {
  // reset() clears startedAt, so the next end-of-game gets a fresh
  // (code, started_at) tuple and archives as a new match row.
  return run(code, (g) => g.reset(playerId));
}

export async function leaveRoom(code: string, playerId: string): Promise<ActionResult> {
  const c = sanitizeCode(code);
  const game = gameStore.get(c);
  if (!game) return ok();
  game.leave(playerId);
  // leave() can flip the game to 'ended' (last-alive-standing); cover
  // that case before we possibly drop the room.
  persist(game);
  game.notify();
  // gameStore.cleanup also deletes the persisted row when empty.
  gameStore.cleanup(c);
  return ok();
}

File: backend/src/routes/actions.rs
//! 动作接口——对应原 Server Actions 的每个动词。
//!
//! 一律 POST + JSON body，body 里都带 `playerId`，路径里都带
//! `code`。响应统一走 `ActionResponse` 信封。
//!
//! 加入 / 重开两个动作不能直接走 `Store::mutate`：前者要在房间不
//! 存在时新建，后者顺路要清下"已归档"标记。剩下六个全部委托给
//! `mutate` 自动 persist + notify。

use super::{ActionResponse, SharedState};
use crate::game::GameError;
use crate::store::{Room, Store};
use crate::types::{
    ContinueRequest, DrawRequest, GuessRequest, JoinRequest, PlaceJokerRequest, PlayerIdOnly,
    Phase,
};
use axum::extract::{Path, State};
use axum::Json;
use std::sync::Arc;

fn from_err(e: GameError) -> ActionResponse {
    ActionResponse::err(e.to_string())
}

pub async fn join_or_create(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<JoinRequest>,
) -> Json<ActionResponse> {
    if raw_code.is_empty() || req.player_id.is_empty() || req.name.is_empty() {
        return Json(ActionResponse::err("请提供房间码、玩家与昵称"));
    }
    let code = Store::sanitize_code(&raw_code);

    // 三种情况：房间不存在、已是房间成员、想新加入
    let room = match state.store.get(&code) {
        // 1. 全新房间——第一个进来的人就是 host
        None => {
            let room = Arc::new(Room::new(code.clone(), req.player_id.clone()));
            room.state.lock().add_player(&req.player_id, &req.name);
            state.store.insert(room.clone());
            room
        }
        Some(room) => {
            let already_in = room
                .state
                .lock()
                .players
                .iter()
                .any(|p| p.id == req.player_id);

            if already_in {
                // 2. 重连场景，addPlayer 自己处理 connected/alive 回滚
                room.state.lock().add_player(&req.player_id, &req.name);
                state.store.persist_and_maybe_archive(&room);
                room.notify();
                return Json(ActionResponse::ok());
            }

            // 3. 新玩家加入既有房间——先把准入条件验完
            {
                let g = room.state.lock();
                if req.as_host {
                    return Json(from_err(GameError::RoomExists));
                }
                if !matches!(g.phase, Phase::Waiting) {
                    return Json(from_err(GameError::GameInProgress));
                }
                if g.players.len() >= 4 {
                    return Json(from_err(GameError::RoomFull));
                }
                if g.has_player_name(&req.name) {
                    return Json(from_err(GameError::NameTaken));
                }
            }
            room.state.lock().add_player(&req.player_id, &req.name);
            room
        }
    };

    state.store.persist_and_maybe_archive(&room);
    room.notify();
    Json(ActionResponse::ok())
}

pub async fn start_game(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<PlayerIdOnly>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    match state.store.mutate(&code, |g| g.start(&req.player_id)) {
        Ok(_) => Json(ActionResponse::ok()),
        Err(e) => Json(from_err(e)),
    }
}

pub async fn draw_tile(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<DrawRequest>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    match state.store.mutate(&code, |g| g.draw_tile(&req.player_id, req.color)) {
        Ok(_) => Json(ActionResponse::ok()),
        Err(e) => Json(from_err(e)),
    }
}

pub async fn guess_tile(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<GuessRequest>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    let Some(room) = state.store.get(&code) else {
        return Json(from_err(GameError::RoomNotFound));
    };
    // guess() 返回 reveal 信息，要在 notify 之前先单独广播一次
    // ——前端动画依赖独立的 reveal 事件，不是从 public state 里
    // 反推差异。所以这里没走 mutate。
    let reveal = {
        let mut g = room.state.lock();
        g.guess(&req.player_id, &req.target_player_id, &req.tile_id, req.number)
    };
    match reveal {
        Ok(reveal) => {
            room.emit_reveal(&reveal);
            state.store.persist_and_maybe_archive(&room);
            room.notify();
            Json(ActionResponse::ok())
        }
        Err(e) => Json(from_err(e)),
    }
}

pub async fn place_joker(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<PlaceJokerRequest>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    match state
        .store
        .mutate(&code, |g| g.place_joker(&req.player_id, req.position))
    {
        Ok(_) => Json(ActionResponse::ok()),
        Err(e) => Json(from_err(e)),
    }
}

pub async fn decide_continue(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<ContinueRequest>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    match state
        .store
        .mutate(&code, |g| g.decide_continue(&req.player_id, req.cont))
    {
        Ok(_) => Json(ActionResponse::ok()),
        Err(e) => Json(from_err(e)),
    }
}

pub async fn reset_game(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<PlayerIdOnly>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    match state.store.mutate(&code, |g| g.reset(&req.player_id)) {
        Ok(_) => Json(ActionResponse::ok()),
        Err(e) => Json(from_err(e)),
    }
}

pub async fn leave_room(
    State(state): State<SharedState>,
    Path(raw_code): Path<String>,
    Json(req): Json<PlayerIdOnly>,
) -> Json<ActionResponse> {
    let code = Store::sanitize_code(&raw_code);
    let Some(room) = state.store.get(&code) else {
        // 房间已经清掉的话，leave 是 no-op，但别返回错——前端
        // 不在乎区分这两种情况
        return Json(ActionResponse::ok());
    };
    room.state.lock().leave(&req.player_id);
    state.store.persist_and_maybe_archive(&room);
    room.notify();
    state.store.cleanup(&code);
    Json(ActionResponse::ok())
}
